### PR TITLE
cache: add statscache metrics and enable LFU

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: ref #45367

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/cache/internal/lfu/lfu_cache.go [hunk 1]
```diff
@@ -130,3 +148,9 @@ func (s *LFU) wait() {
 func (s *LFU) metrics() *ristretto.Metrics {
 	return s.cache.Metrics
 }
+
+// Stop implements statsCacheInner
+func (s *LFU) Stop() {
```
Comment (hawkingrei) [inline statistics/handle/cache/internal/lfu/lfu_cache.go:153]: Add ```stop``` to avoid leak.

File: statistics/handle/cache/statscacheinner.go [hunk 2]
```diff
@@ -105,10 +121,29 @@ func (sc *StatsCache) PutFromInternal(id int64, t *statistics.Table) {
...
+	return ok
+}
+
 // Put puts the table statistics to the cache.
 func (sc *StatsCache) put(id int64, t *statistics.Table, moveLRUFront bool) {
-	sc.c.Put(id, t, moveLRUFront)
-
+	ok := sc.putCache(id, t, moveLRUFront)
+	if !ok {
+		logutil.BgLogger().Warn("fail to put the stats cache")
```
Comment (time-and-fate) [inline statistics/handle/cache/statscacheinner.go:144]: I suggest we make the log content more detailed here, like adding the table name, stack trace, or anything useful.

File: statistics/handle/cache/internal/lfu/lfu_cache.go [hunk 3]
```diff
@@ -32,6 +34,14 @@ type LFU struct {
 
 // NewLFU creates a new LFU cache.
 func NewLFU(totalMemCost int64) (*LFU, error) {
+	if totalMemCost == 0 {
+		memTotal, err := memory.MemTotal()
+		if err != nil {
+			return nil, err
+		}
+		totalMemCost = int64(memTotal / 2)
```
Comment (time-and-fate) [inline statistics/handle/cache/internal/lfu/lfu_cache.go:42]: Looks like this is a behavior change, is there any rationale behind this change?

File: statistics/handle/cache/statscacheinner.go [hunk 4]
```diff
@@ -129,9 +164,15 @@ func (sc *StatsCache) Cost() int64 {
 
 // SetCapacity sets the memory capacity of the cache.
 func (sc *StatsCache) SetCapacity(c int64) {
+	metrics.CostGauge.Set(float64(c))
```
Comment (time-and-fate) [inline statistics/handle/cache/statscacheinner.go:167]: We updated this both here and in `(*LFU).SetCapacity()`. I think this is unintended.

File: statistics/handle/handle.go [hunk 5]
```diff
@@ -1012,7 +1022,6 @@ func (h *Handle) statsCacheLen() int {
 // Callers should add retry loop if necessary.
 func (h *Handle) updateStatsCache(newCache *cache.StatsCache) (updated bool) {
 	h.statsCache.Replace(newCache)
-	handle_metrics.CostGauge.Set(float64(newCache.Cost()))
```
Comment (time-and-fate) [inline statistics/handle/handle.go:1015]: Does this mean now we no longer update the metrics when the stats cache updates?

File: statistics/handle/cache/internal/lfu/lfu_cache.go [hunk 6]
```diff
@@ -130,3 +148,9 @@ func (s *LFU) wait() {
 func (s *LFU) metrics() *ristretto.Metrics {
 	return s.cache.Metrics
 }
+
+// Stop implements statsCacheInner
+func (s *LFU) Stop() {
+	s.cache.Close()
+	s.cache.Wait()
```
Comment (time-and-fate) [inline statistics/handle/cache/internal/lfu/lfu_cache.go:155]: Would this possibly wait for a long time?

File: statistics/handle/handle.go (hunk 5 above)
Comment (hawkingrei) [reply statistics/handle/handle.go:1015]: we do not use to update the stats cache to replace the old with the old one. we just set the cache by mutex.

File: statistics/handle/cache/internal/lfu/lfu_cache.go (hunk 6 above)
Comment (hawkingrei) [reply statistics/handle/cache/internal/lfu/lfu_cache.go:155]: No, the goroutine in the cache will receive the close signal and then exit.

File: statistics/handle/cache/internal/inner.go [hunk 7]
```diff
@@ -26,7 +26,7 @@ type StatsCacheInner interface {
 	// Get gets the cache.
 	Get(tid int64, moveFront bool) (*statistics.Table, bool)
 	// Put puts a cache.
-	Put(tid int64, tbl *statistics.Table, moveLRUFront bool)
+	Put(tid int64, tbl *statistics.Table, moveLRUFront bool) bool
```
Comment (qw4990) [inline statistics/handle/cache/internal/inner.go:29]: In our scenarios, it seems that we don't need to expose the `bool`.

File: statistics/handle/cache/internal/lfu/lfu_cache.go (hunk 1 above)
Comment (qw4990) [inline statistics/handle/cache/internal/lfu/lfu_cache.go:153]: How about renaming it from `Stop()` to `Close()` to keep aligned with the underlying `s.cache.Close()`?

File: statistics/handle/cache/internal/lfu/lfu_cache.go (hunk 6 above)
Comment (qw4990) [reply statistics/handle/cache/internal/lfu/lfu_cache.go:155]: Will it cause some dead-lock or dead-goroutine if we call `cache.Close()` before `cache.Wait()`?

File: domain/domain.go [hunk 8]
```diff
@@ -1022,6 +1022,9 @@ func (do *Domain) Close() {
 	if intest.InTest {
 		infosync.MockGlobalServerInfoManagerEntry.Close()
 	}
+	if handle := do.statsHandle.Load(); handle != nil {
```
Comment (qw4990) [inline domain/domain.go:1025]: Is this change necessary? If it is not necessary for this PR, I prefer to keep it unchanged to avoid any risk 🤦

File: statistics/handle/cache/internal/lru/lru_cache.go [hunk 9]
```diff
@@ -1,443 +0,0 @@
-// Copyright 2023 PingCAP, Inc.
```
Comment (qw4990) [inline statistics/handle/cache/internal/lru/lru_cache.go:1]: Better to keep it unchanged, we can remove it in the next few PRs after finishing the new LFU completely.

File: statistics/handle/cache/statscacheinner.go (hunk 4 above)
Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go:167]: If we read the code from the coprocessor cache, we will know the capacity is from the cost and the user can config the max memory usage of the coprocessor. this LFU is the same approach.

Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go:167]: the function can be changed as ```SetCost```.

File: statistics/handle/cache/statscacheinner.go [hunk 10]
```diff
@@ -129,9 +164,15 @@ func (sc *StatsCache) Cost() int64 {
...
 	sc.c.SetCapacity(c)
 }
 
+// Stop stops the cache.
+func (sc *StatsCache) Stop() {
```
Comment (qw4990) [inline statistics/handle/cache/statscacheinner.go:172]: Better to use `Close` to aligned with the underlying cache's method name.
And in addition, `Close` seems a better name in Golang than `Stop`, many standard libs use `Close` instead of `Stop` like `io.Closer`.

File: statistics/handle/cache/statscacheinner.go [hunk 11]
```diff
@@ -57,17 +59,21 @@ func WithTableStatsByQuery() TableStatsOpt {
...
 
 // NewStatsCache creates a new StatsCacheWrapper.
-func NewStatsCache() *StatsCache {
+func NewStatsCache() (*StatsCache, error) {
 	enableQuota := config.GetGlobalConfig().Performance.EnableStatsCacheMemQuota
 	if enableQuota {
 		capacity := variable.StatsCacheMemQuota.Load()
-		return &StatsCache{
-			c: lru.NewStatsLruCache(capacity),
+		stats, err := lfu.NewLFU(capacity)
```
Comment (qw4990) [inline statistics/handle/cache/statscacheinner.go:66]: In method `CopyAndUpdate`, the `StatsCache` copies the underlying `cache`, will that cause any risk since the LFU cache doesn't implement `copy` method?

File: statistics/handle/cache/internal/lfu/lfu_cache.go (hunk 1 above)
Comment (hawkingrei) [reply statistics/handle/cache/internal/lfu/lfu_cache.go]: Done

File: domain/domain.go (hunk 8 above)
Comment (hawkingrei) [reply domain/domain.go:1025]: It is to avoid the ```goleak```. so that we do not add ```ristretto``` into goleak config in the every ```main_test```.

File: statistics/handle/cache/internal/lfu/lfu_cache.go (hunk 6 above)
Comment (hawkingrei) [reply statistics/handle/cache/internal/lfu/lfu_cache.go:155]: Let's start by looking at the implementation of ```wait``` and ```close```. 

- ```Close``` is at the https://github.com/dgraph-io/ristretto/blob/main/cache.go#L354.

Close is to close some channels to send a close signal. so that the background task will be stopped like evict policy. Finally, it will be marked as closed.

- ```Wait``` is at the https://github.com/dgraph-io/ristretto/blob/main/cache.go#L213

 It seems that "wait" doesn't have any effect because once you close it, it returns immediately.

Comment (hawkingrei) [reply statistics/handle/cache/internal/lfu/lfu_cache.go:155]: it has been marked as closed. All the ```get```/```set``` will return nothing immediately.

File: statistics/handle/cache/internal/inner.go (hunk 7 above)
Comment (hawkingrei) [reply statistics/handle/cache/internal/inner.go:29]: Actually, in the ```StatsCache```, the retry logic is already included, so the internal cache needs to expose its status for ```StatsCache``` to handle retries and logging.

File: statistics/handle/cache/internal/lru/lru_cache.go (hunk 9 above)
Comment (hawkingrei) [reply statistics/handle/cache/internal/lru/lru_cache.go]: Done, it has been recovered.

File: statistics/handle/cache/statscacheinner.go (hunk 10 above)
Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go]: Done

File: statistics/handle/cache/statscacheinner.go (hunk 11 above)
Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go:67]:
```
h.updateStatsCache(oldCache.CopyAndUpdate(tables, deletedTableIDs, opts...))
```

I want to merge ```updateStatsCache``` and ```CopyAndUpdate``` into a function. if users use ```LFU```. we do not use COW. But this change will be big. so I will do it in the next PR.

Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go:67]: It is at https://github.com/pingcap/tidb/pull/45518.

File: statistics/handle/cache/statscacheinner.go (hunk 2 above)
Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go:146]: Done

File: statistics/handle/cache/internal/lfu/lfu_cache.go (hunk 3 above)
Comment (hawkingrei) [reply statistics/handle/cache/internal/lfu/lfu_cache.go:42]: In the old LRU, If you set capacity 0, you will have unlimited memory in the cache. It is dangerous.

File: planner/core/plan_stats_test.go [hunk 12]
```diff
@@ -313,7 +313,6 @@ func TestPlanStatsStatusRecord(t *testing.T) {
 	// drop stats in order to change status
 	domain.GetDomain(tk.Session()).StatsHandle().SetStatsCacheCapacity(1)
 	tk.MustQuery("select * from t where b >= 1")
-	require.Equal(t, tk.Session().GetSessionVars().StmtCtx.RecordedStatsLoadStatusCnt(), 2)
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply planner/core/plan_stats_test.go:316]: LFU cannot delete LRU after decreasing the capacity as soon as possible.

[the remark this answers is not in the document]
Comment (hawkingrei) [reply planner/core/plan_stats_test.go:316]: the implementation of ristretto is to evict the cache when the user writes cache.

File: statistics/handle/cache/statscacheinner.go (hunk 11 above)
Comment (hawkingrei) [reply statistics/handle/cache/statscacheinner.go:67]: Done

File: statistics/handle/handle.go (hunk 5 above)
Comment (hawkingrei) [reply statistics/handle/handle.go]: It mean current actual size.

**Review by time-and-fate — APPROVED** [review]

**Comment by hawkingrei** [comment]
@qw4990 PTAL

File: statistics/handle/cache/statscacheinner.go [hunk 13]
```diff
@@ -105,10 +122,30 @@ func (sc *StatsCache) PutFromInternal(id int64, t *statistics.Table) {
...
+		return ok
+	}
+	// retry three times and sleep exponentially. 20ms, 40ms and 80ms.
+	for i := 0; i < 3; i++ {
+		time.Sleep(time.Duration(1<<uint(i)) * 10 * time.Millisecond)
+		ok = sc.c.Put(id, t, moveLRUFront)
+		if ok {
+			return ok
+		}
+	}
```
Comment (qw4990) [inline statistics/handle/cache/statscacheinner.go:137]: Please don't do that... this retry seems very dangerous, if the `StatsHandler` call this, in the worst case, `StatsHandler` will be blocked for `140ms`.
I suggest that we just log a warning here and do nothing about retry.

**Review by qw4990 — APPROVED** [review]
